client/src/components/MarkdownTextArea/useAutoResizingTextArea.ts

```diff
@@ -0,0 +1,49 @@
+import { useLayoutEffect, useRef, useState } from 'react';
+import { Maybe } from '../../types';
+
+interface AutoResizingTextArea {
+  handleFocus: () => void;
+  handleBlur: () => void;
+  textAreaRef: React.RefObject<HTMLTextAreaElement>;
+  willRenderMarkdown: boolean;
+}
+
+export function useAutoResizingTextArea(
+  value: string,
+  shouldRenderMarkdown: boolean,
+  onBlurHandler: Maybe<() => void>,
+  extraMargin = '1em'
+): AutoResizingTextArea {
+  const [isFocused, setIsFocused] = useState(false);
+  const textAreaRef = useRef<HTMLTextAreaElement>(null);
+
+  const handleFocus = () => setIsFocused(true);
+  const handleBlur = () => {
+    // Call our custom onBlurHandler
+    if (onBlurHandler) {
+      onBlurHandler();
+    }
+
+    // Call our unfocus after, to switch back to the markdown render
+    setIsFocused(false);
+  };
+
+  const willRenderMarkdown = !isFocused && !!value && shouldRenderMarkdown;
+
+  useLayoutEffect(() => {
```

Comment:
this is the real bread and butter:
1. Use `useLayoutEffect` to make sure updates to the DOM are made before React paints the DOM
2. Use the linked SO solution to get a (seemingly) perfect height calculation.
3. As a bonus, jump the cursor to the end